src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp

```diff
@@ -0,0 +1,199 @@
+#include "Initializer/typedefs.hpp"
+#ifdef ACL_DEVICE
+
+#include <Parallel/MPI.h>
+#include "Solver/time_stepping/GhostTimeClusterWithCopy.h"
+#include "device.h"
+
+namespace seissol::time_stepping {
+template <MPI::DataTransferMode CommType>
+GhostTimeClusterWithCopy<CommType>::GhostTimeClusterWithCopy(
+    double maxTimeStepSize,
+    int timeStepRate,
+    int globalTimeClusterId,
+    int otherGlobalTimeClusterId,
+    const MeshStructure* meshStructure)
+    : AbstractGhostTimeCluster(maxTimeStepSize,
+                               timeStepRate,
+                               globalTimeClusterId,
+                               otherGlobalTimeClusterId,
+                               meshStructure) {
+  numberOfRegions = meshStructure->numberOfRegions;
+  prefetchCopyRegionsStreams.resize(numberOfRegions);
+  prefetchGhostRegionsStreams.resize(numberOfRegions);
+  receiveRegionsStates.resize(numberOfRegions);
+
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    prefetchCopyRegionsStreams[region] = device.api->createGenericStream();
+    prefetchGhostRegionsStreams[region] = device.api->createGenericStream();
+    receiveRegionsStates[region] = ReceiveState::RequiresMpiTesting;
+  }
+
+  duplicatedCopyRegions.resize(numberOfRegions);
+  duplicatedGhostRegions.resize(numberOfRegions);
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    const size_t copyRegionSize = meshStructure->copyRegionSizes[region] * sizeof(real);
+    const size_t ghostRegionSize = meshStructure->ghostRegionSizes[region] * sizeof(real);
+    if constexpr (CommType == MPI::DataTransferMode::CopyInCopyOutDevice) {
+      duplicatedCopyRegions[region] = static_cast<real*>(device.api->allocGlobMem(copyRegionSize));
+      duplicatedGhostRegions[region] =
+          static_cast<real*>(device.api->allocGlobMem(ghostRegionSize));
+    } else {
+      duplicatedCopyRegions[region] =
+          static_cast<real*>(device.api->allocPinnedMem(copyRegionSize));
+      duplicatedGhostRegions[region] =
+          static_cast<real*>(device.api->allocPinnedMem(ghostRegionSize));
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+GhostTimeClusterWithCopy<CommType>::~GhostTimeClusterWithCopy() {
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    device.api->destroyGenericStream(prefetchCopyRegionsStreams[region]);
+    device.api->destroyGenericStream(prefetchGhostRegionsStreams[region]);
+    if constexpr (CommType == MPI::DataTransferMode::CopyInCopyOutDevice) {
+      device.api->freeMem(duplicatedCopyRegions[region]);
+      device.api->freeMem(duplicatedGhostRegions[region]);
+    } else {
+      device.api->freePinnedMem(duplicatedCopyRegions[region]);
+      device.api->freePinnedMem(duplicatedGhostRegions[region]);
+    }
+  }
+}
+
```

Comment:
If you have a custom dtor you should also have custom copy / move operators and copy-/move-initializers. If you do not plan on copying / moving objects of this class then delete the 4 operators.

Reply:
I deleted them